### ISSUE TITLE
代码中的value loss和 critic loss 是不是有错误

### ISSUE THREAD

**Issue by caowanxia123** [THUDM/WebRL#39]
#![Image](https://github.com/user-attachments/assets/498fc7bf-8f16-4a31-8fe3-1c3ee60ad839)

请问一下这里在外面求了softmax再使用torch.nn.CrossEntropyLoss()是不是有问题？

critic loss 中计算advantage 似乎也有问题？
![Image](https://github.com/user-attachments/assets/309ef735-fa85-4315-8710-35ebd71e15ee)/
advantage 这里的0.05是怎么来的，似乎和论文也对不上？

`def actor_loss(self, observation, action, next_observation, mc_return, pi_action, advantage, reward, add_sft_loss=False,
                   validation=False,**kwargs):
        mc_return = torch.Tensor(mc_return).to(self.accelerator[1].unwrap_model(self.agent.model).device, dtype = self.accelerator[1].unwrap_model(self.agent.model).dtype).flatten()
        reward = torch.Tensor(reward).to(self.accelerator[1].unwrap_model(self.agent.model).device, dtype = self.accelerator[1].unwrap_model(self.agent.model).dtype).flatten()
        with torch.no_grad():
            v1, v2 = self.agent.critic(observation, action, detach_model=False)
            nv1, nv2 = self.agent.critic(next_observation, action, detach_model=False)
            v1 = self.softmax(v1)[:, 1]
            v2 = self.softmax(v2)[:, 1]
            nv1 = self.softmax(nv1)[:, 1]
            nv2 = self.softmax(nv2)[:, 1]
            v = torch.minimum(v1, v2).flatten()
            nv = torch.minimum(nv1, nv2).flatten()
            advantage = nv - v - 0.05 + reward + mc_return
            advantage = torch.clamp(advantage, -1, 1).to(self.accelerator[1].unwrap_model(self.agent.model).device, dtype = self.accelerator[1].unwrap_model(self.agent.model).dtype)

        log_prob, action_attention_mask = self.agent.get_log_prob(observation, action)
        log_prob_detach = log_prob.detach()
        ref_log_prob = self.agent.get_log_prob_ref(observation, action).to(log_prob.device)
        non_zero_counts = action_attention_mask.sum(dim=1)
        
        advantages = advantage.flatten()
        advantages = torch.where(advantages >= 0, torch.tensor(1.0), torch.tensor(-0.6)).to(log_prob.device)
        
        log_prob = log_prob.sum(dim = 1).flatten()
        ref_log_prob = ref_log_prob.sum(dim = 1).flatten()
        # ref_prob = torch.exp(ref_log_prob).mean(dim = 1)
        ref_prob = torch.exp(ref_log_prob / non_zero_counts)
        
        beta = torch.ones_like(advantages).to(advantages.device)
        cond1 = (advantages >= 0) & (ref_prob >= 0.8)
        cond2 = (advantages < 0) & (ref_prob < 0.9)
        beta[cond1] = 5.0
        beta[cond2] = 5.0
        
        # loss = self.mse_loss(beta * log_prob_ratio, advantages)
        mask = (advantages > 0).to(dtype=log_prob.dtype, device=log_prob.device)
        
        non_zero_counts_mask = non_zero_counts[mask.bool()].sum()
        safe_count = torch.where(non_zero_counts_mask > 0, non_zero_counts_mask, torch.tensor(1.0, dtype=non_zero_counts_mask.dtype))
        
        log_prob_ratio = (log_prob - ref_log_prob) / safe_count
        
        ratio = torch.abs(advantages / log_prob_ratio)
        beta = torch.where(beta <= ratio, beta, ratio).detach()
        
        loss = self.mse_loss(beta * log_prob_ratio, advantages)
        loss = torch.reciprocal(beta) * loss * mask
        # loss = loss * mask
        loss = loss.mean()
        value_loss = torch.zeros_like(loss)
        if add_sft_loss == True:
            log_prob_detached = log_prob.detach()
            prob_detached = torch.exp(log_prob_detached / non_zero_counts)
            coefficient = torch.exp(1 - prob_detached)
            
            sft_loss = -(log_prob * coefficient)[mask.bool()].sum() / safe_count
            # sft_loss = -log_prob[mask.bool()].sum() / safe_count
            loss += sft_loss
            
        # if not validation and loss is not None and mask.sum() != 0:
        if not validation and loss is not None:
            self.accelerator[1].backward(loss + value_loss)

        advantages = advantages.detach().cpu()
        info =  {"pg.loss": loss.detach().cpu().item() if loss is not None else 0,
                "advantages.mean": advantages.mean(),
                "log_prob": (log_prob / safe_count).cpu().mean(),
                "ref_prob": ref_prob.cpu().mean(),
                "ref_log_prob": (ref_log_prob / safe_count).cpu().mean(),
                "log_prob_ratio": log_prob_ratio.cpu().mean(),
                "mask": mask.cpu().sum(),
                "beta": beta.cpu().mean(),
                "sft_loss": sft_loss.cpu().mean() if add_sft_loss == True else 0,
                "non_zero_counts": torch.sum(non_zero_counts).cpu(),
                "beta_log_prob_ratio": (beta * log_prob_ratio).cpu().mean()}
        if validation:
            validation_info = {}
            for k,v in info.items():
                validation_info["validation."+k] = v
            validation_info['log_prob'] = log_prob.detach().cpu().mean()
            return validation_info
        return info`

`advantages = advantage.flatten()
        advantages = torch.where(advantages >= 0, torch.tensor(1.0), torch.tensor(-0.6)).to(log_prob.device)`这里是为什么，为什么大于0的取1，小于0的取-0.6这是怎么来地

**Comment by QZH-777** [comment]
感谢您的提醒！关于softmax部分，确实应该删去。原始的训练代码没有softmax的操作，但在整理到仓库时出了一点问题。关于advantage部分的0.05，这里我们采取了和DigiRL一样的操作。在论文中我们也提到了关于advantage采用的是DigiRL的方法，因此就采用了DigiRL这部分的实现代码。关于advantage的放缩，我们在实验中发现通过这种方式可以使训练更稳定。

**Comment by caowanxia123** [comment]
> 感谢您的提醒！关于softmax部分，确实应该删去。原始的训练代码没有softmax的操作，但在整理到仓库时出了一点问题。关于advantage部分的0.05，这里我们采取了和DigiRL一样的操作。在论文中我们也提到了关于advantage采用的是DigiRL的方法，因此就采用了DigiRL这部分的实现代码。关于advantage的放缩，我们在实验中发现通过这种方式可以使训练更稳定。

我在修改了softmax部分的代码后进行了复现，advantages = advantage.flatten() advantages = torch.where(advantages >= 0, torch.tensor(1.0), 在使用这个放缩时pg.loss无法收敛，将这个放缩去掉，并将beta固定且设为0.1~1,pg.loss可以收敛，但是phase1的success rate，相比于sft下降，请问一下开源的代码中的参数是否是每个phase训练都用的参数，还是每个phase参数会变化？

**Comment by swt-user** [comment]
请问这个问题有进一步进展吗？我目前也是复现phase 1之后success rate 下降
